upload_run.py

```diff
@@ -69,14 +37,119 @@ def __call__(self, parser, namespace, values, option_string=None):
                                   'The list of runs is generated at start time '
                                   '(Runs added to directory mid upload will not be uploaded).')
 
+# Optional arguments for overriding config file settings
+# Explanation:
+#   nargs='?', const=True, default=False,
+#       Allows zero or one parameters
+#       If the argument is not given:                      the value will be False           (indicates load from file)
+#       If the argument is given, and no parameter given:  the value will be True            (prompt user for input)
+#       If the argument is given, and parameter is given:  the value will be the parameter   (used to override config)
+argument_parser.add_argument('-ci', '--config_client_id', action='store', nargs='?', const=True, default=False,
+                             help='Override the "client_id" field in config file. '
+                                  'This is for the uploader client created in IRIDA, used to handle the data upload')
+argument_parser.add_argument('-cs', '--config_client_secret', action='store', nargs='?', const=True, default=False,
+                             help='Override "client_secret" in config file. '
+                                  'This is for the uploader client created in IRIDA, used to handle the data upload')
+argument_parser.add_argument('-cu', '--config_username', action='store', nargs='?', const=True, default=False,
+                             help='Override "username" in config file. This is your IRIDA account username.')
+argument_parser.add_argument('-cp', '--config_password', action='store', nargs='?', const=True, default=False,
+                             help='Override "password" in config file. This corresponds to your IRIDA account.')
+argument_parser.add_argument('-cb', '--config_base_url', action='store', nargs='?', const=True, default=False,
+                             help='Override "base_url" in config file. The api url for your irida instance '
+                                  '(example: https://my.irida.server/api/)')
+argument_parser.add_argument('-cr', '--config_parser', action='store', nargs='?', const=True, default=False,
+                             help='Override "parser" in config file. '
+                                  'Choose the type of sequencer data that is being uploaded. '
+                                  'Supported parsers: ' + str(supported_parsers))
+
+
+def _set_config_override(args):
+    """
+    Check the config override arguments and override
+    :param args: list of args from parseargs
+    :return:
+    """
+    client_id = None
+    client_secret = None
+    username = None
+    password = None
+    base_url = None
+    parser = None
+
+    if args.config_client_id is True:
```

Comment:
An issue happens with the blank option=prompt thing.  If try to run this when the last argument is a prompt, it fails since it captures the directory name as the value for the prompted option.  Example: 
```bash
$ ./irida-uploader.sh -cb http://somewhere/api -ci client -cs secret -cu user -cp /files/to/upload 
usage: upload_run.py [-h] [--version] [-c CONFIG] [-f] [-b]
                     [-ci [CONFIG_CLIENT_ID]] [-cs [CONFIG_CLIENT_SECRET]]
                     [-cu [CONFIG_USERNAME]] [-cp [CONFIG_PASSWORD]]
                     [-cb [CONFIG_BASE_URL]] [-cr [CONFIG_PARSER]]
                     directory
upload_run.py: error: the following arguments are required: directory
```
Note the `-cp` at the end set to prompt for password.

One thing you could do is add a `-d` arg for the directory.  I don't mind that since then it makes everything consistent.

Reply:
Another option could be to change the way the overrides work.
1. if user adds the option on the command line, use it.
2. if not check the config file. if it's there use it
3. if neither 1 or 2, prompt.

This removes the blank option=prompt thing but that's not a big deal to me.  The only other program I can think of that works that way is mysql.

Reply:
I fixed it by changing directory to be an optional argument, then setting it to required. This made it show up as an optional argument so I suppressed it across the help statements and added the information about the argument to the main ArgumentParser description and prog fields.

This is not the prettiest code, but it suits our use case pretty well.

Reply:
I'd say not too hacky.  It serves the purpose.  It was annoying that it showed `-d` as an "optional", so suppressing the auto help message for it works just fine.  Thanks.